lib/utils/apiUtils.js

```diff
@@ -59,104 +65,219 @@ async function refreshTokens(firmId) {
-  if (!firmTokens) {
-    consola.error(`Missing authorization for firm id: ${firmId}`);
-    process.exit(1);
+function setAxiosDefaults(type, envId) {
+  let axiosInstance;
+
+  if (type == "firm") {
+    const firmTokens = firmCredentials.getTokenPair(envId);
+    if (!firmTokens) {
+      consola.error(`Missing authorization for firm id: ${envId}`);
+      process.exit(1);
+    }
+
+    axiosInstance = axios.create({
+      baseURL: `${BASE_URL}/api/v4/f/${envId}`,
+      headers: {
+        "User-Agent": `silverfin-cli/${pkg.version}`,
+        Authorization: `Bearer ${firmTokens.accessToken}`,
+      },
+    });
+
+    axiosInstance.interceptors.response.use(
+      (res) => {
+        return res;
+      },
+      async (error) => {
+        const originalConfig = error.config;
+        if (error.response) {
+          if (error.response.status === 401 && !originalConfig._retry) {
+            // Set _retry to true after trying the first refresh request on 401 status to avoid an infinite loop
+            originalConfig._retry = true;
+
+            try {
+              // Get a refreshed set of tokens
+              const firmId = originalConfig.baseURL.split("/").pop();
+              await refreshTokens(axiosInstance, firmId);
+
+              // Set the Authorization header with the new access token
+              const firmTokens = firmCredentials.getTokenPair(firmId);
+              consola.debug(`refreshed tokens for firm ${firmId}`);
+
+              axiosInstance.defaults.headers.common[
+                "Authorization"
+              ] = `Bearer ${firmTokens.accessToken}`;
+
+              originalConfig.headers.Authorization = `Bearer ${firmTokens.accessToken}`;
+
+              return axiosInstance(originalConfig);
+            } catch (_error) {
+              consola.error(
+                `Error 401: Failed to refresh the firm access token automatically, try to manually authorize the firm again with the authorize command`
+              );
+
+              if (_error.response && _error.response.data) {
+                return Promise.reject(_error.response.data);
+              }
+
+              return Promise.reject(_error);
+            }
+          }
+        }
+
+        return Promise.reject(error);
+      }
+    );
+  } else if (type == "partner") {
+    axiosInstance = axios.create({
+      baseURL: `${BASE_URL}/api/partner/v1`,
+      headers: {
+        "User-Agent": `silverfin-cli/${pkg.version}`,
+      },
+    });
+
+    axiosInstance.interceptors.request.use((config) => {
+      // Fetch the stored partner api key
+      const partnerToken = firmCredentials.getPartnerCredentials(envId)?.token;
+
+      // Set the partner api key and id as a query param to every request
+      config.params = {
+        ...config.params,
+        partner_id: envId,
+        api_key: partnerToken,
+      };
+
+      return config;
+    });
+
+    axiosInstance.interceptors.response.use(
```

Comment:
Just to confirm: this is the way you'd add an error handler to the axios instance, right? If yes, mind adding a comment above explaining this as it doesn't look intuitive in the 1st look.

Reply:
We generally still use the `function responseErrorHandler` in the `apiUtils` file for error handling of the response. 

At the moment, the interceptor will only handle the "401 unauthorized" error to handle the refreshing of the tokens. 

I left a complete refactoring of our error handling with interceptors out of scope of this update.